accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between and (2) are disjoint (such as iOS and Windows), the
+  information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
+2. '{API}' has been deprecated since {OS} {OS version}.
+3. '{API}' has been removed since {OS} {OS version}.
+
+### Automatic suppression via platform guards
+
+Diagnostic (1) can be suppressed via a call to
+`RuntimeInformation.IsOSPlatformOrLater()`. This should work for for simple cases
+where the call is contained inside an `if` check but also when the check causes
+control flow to never reach the call:
+
+```C#
+public void Case1()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        AppleApi();
+}
+
+public void Case2()
+{
+    if (!RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 13))
+        return;
+
+    AppleApi();
+}
+```
+
+Diagnostics (2) and (3) are analogous except the guard is provided by
+`RuntimeInformation.IsOSPlatformAndEarlier()`.
+
+### Code Fixers
+
+* Suggest wrapping statement in platform guard
+* Suggest annotating the method with an attribute
+
+All diagnostics can be suppressed by surround the call site using a platform
```

Comment:
Is this going to be the only way to suppress the message? Are we going to support suppressing it with a #pragma or a `SuppressMessage` attribute?

Reply:
Yes and there is no way to not support this because that's what Roslyn already does for all analyzers. IOW, the analyzers can report whatever it wants and if the user suppressed it globally via `<NoWarn>`, `SuppressMessage`, or via `#prama` the compiler won't report it.